python/cudf_polars/cudf_polars/experimental/parallel.py

```diff
@@ -0,0 +1,152 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Partitioned LogicalPlan nodes."""
+
+from __future__ import annotations
+
+import operator
+from functools import reduce, singledispatch
+from typing import TYPE_CHECKING, Any
+
+from cudf_polars.dsl.ir import IR
+from cudf_polars.dsl.traversal import traversal
+
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+    from typing import TypeAlias
+
+    from cudf_polars.containers import DataFrame
+    from cudf_polars.dsl.nodebase import Node
+    from cudf_polars.typing import GenericTransformer
+
+
+class PartitionInfo:
+    """
+    Partitioning information.
+
+    This class only tracks the partition count (for now).
+    """
+
+    __slots__ = ("count",)
+
+    def __init__(self, count: int):
+        self.count = count
+
+
+LowerIRTransformer: TypeAlias = (
+    "GenericTransformer[IR, MutableMapping[IR, PartitionInfo]]"
+)
+"""Protocol for Lowering IR nodes."""
+
+
+def get_key_name(node: Node) -> str:
+    """Generate the key name for a Node."""
+    return f"{type(node).__name__.lower()}-{hash(node)}"
+
+
+@singledispatch
+def lower_ir_node(
+    ir: IR, rec: LowerIRTransformer
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    """Rewrite an IR node with proper partitioning."""
+    raise AssertionError(f"Unhandled type {type(ir)}")  # pragma: no cover
+
+
+@lower_ir_node.register(IR)
+def _default_lower_ir_node(
```

Comment:
This function **will** get used a lot when we start implementing multi-partition logic (e.g. in https://github.com/rapidsai/cudf/pull/17387), because we will want to "fall back" to this logic when we only have one partition.

Reply:
Yes, but it should be used by calling `lower_ir_node` and registering all things that it handles. Rather than calling the "private" function by hand

Reply:
When we call `lower_ir_node` on a `GroupBy` node, for example, we will need to check if our parallel-groupby logic supports the particular aggregation. If not, we will want to return `_default_lower_ir_node(ir, rec)` (and let that function complain about missing multi-partition support if necessary).

In order to **completely** cover this case with `lower_ir_node` dispatching, we would need to lower the original `GroupBy` object into a new class type that we register to the same logic as `IR`. That is okay with me, but I just wanted to check that this is what you have in mind before I make the changes?

Reply:
Oh, I see. So you want the single-partition lowering to be a fallback option for multi-partition.

And you kind of want to call the "super" method in case of fallback.

However, singledispatch doesn't stack like that.

How about this:

- We implement single-partition lowering as here `lower_single_partition`
- In the multiple partition case, we have a new function `lower_multi_partition`, whose default implementation can just defer to `lower_single_partition`, but for groupby (say) you do something special and can call the fallback.

To have access to all the caching mappers you would do something like (sketch):

```
@singledispatch
def _lower_single(node: IR, rec: ...) -> ...:
  ...

@singledispatch
def _lower_multi(node: IR, rec: ...) -> ...:
    if cant_lower_multi(node):
        return rec.state["default_mapper"](node)
    ...

def lower_multi(node: IR):
    mapper = CachingVisitor(_lower_multi, state={"default_mapper": CachingVisitor(_lower_single)})
    return mapper(node)
```

Reply:
Yes, this seems like a reasonable way to get what we want. I'll plan to incorporate something like this in a simple follow-up PR (that also adds simple partition-wise `DataFrameScan` and `Select` functionality).